File: common/src/main/java/juuxel/adorn/block/PicketFenceBlock.java
package juuxel.adorn.block;

import juuxel.adorn.util.ShapeRotation;
import net.minecraft.world.level.block.Block;
import net.minecraft.world.level.block.state.BlockBehaviour.Properties;
import net.minecraft.world.level.block.state.BlockState;
import net.minecraft.world.phys.shapes.CollisionContext;
import net.minecraft.world.level.block.SimpleWaterloggedBlock;
import net.minecraft.world.level.pathfinder.PathComputationType;
import net.minecraft.world.level.material.FluidState;
import net.minecraft.world.level.material.Fluids;
import net.minecraft.world.item.context.BlockPlaceContext;
import net.minecraft.world.level.block.state.StateDefinition;
import net.minecraft.world.level.block.state.properties.BooleanProperty;
import net.minecraft.world.level.block.state.properties.EnumProperty;
import net.minecraft.world.level.block.state.properties.BlockStateProperties;
import net.minecraft.world.level.block.Mirror;
import net.minecraft.world.level.block.Rotation;
import net.minecraft.util.StringRepresentable;
import net.minecraft.core.BlockPos;
import net.minecraft.core.Direction;
import net.minecraft.util.RandomSource;
import net.minecraft.world.phys.shapes.VoxelShape;
import net.minecraft.world.level.BlockGetter;
import net.minecraft.world.level.LevelReader;
import net.minecraft.world.level.ScheduledTickAccess;

import java.util.Map;

public final class PicketFenceBlock extends Block implements SimpleWaterloggedBlock, BlockWithDescription {
    public static final EnumProperty<Shape> SHAPE = EnumProperty.create("shape", Shape.class);
    public static final EnumProperty<Direction> FACING = BlockStateProperties.HORIZONTAL_FACING;
    public static final BooleanProperty WATERLOGGED = BlockStateProperties.WATERLOGGED;

    private static final Map<Direction, VoxelShape> STRAIGHT_OUTLINE_SHAPES = ShapeRotation.buildShapeRotationsFromNorth(0, 0, 7, 16, 16, 9);
    private static final Map<Direction, VoxelShape> CORNER_OUTLINE_SHAPES = ShapeRotation.mergeIntoShapeMap(
        ShapeRotation.mergeShapeMaps(
            ShapeRotation.buildShapeRotationsFromNorth(0, 0, 7, 9, 16, 9),
            ShapeRotation.buildShapeRotationsFromNorth(7, 0, 9, 9, 16, 16)
        ),
        PostBlock.Y_SHAPE
    );
    private static final Map<Direction, VoxelShape> STRAIGHT_COLLISION_SHAPES = ShapeRotation.buildShapeRotationsFromNorth(0, 0, 7, 16, 24, 9);
    private static final Map<Direction, VoxelShape> CORNER_COLLISION_SHAPES = ShapeRotation.mergeIntoShapeMap(
        ShapeRotation.mergeShapeMaps(
            ShapeRotation.buildShapeRotationsFromNorth(0, 0, 7, 9, 24, 9),
            ShapeRotation.buildShapeRotationsFromNorth(7, 0, 9, 9, 24, 16)
        ),
        box(6.0, 0.0, 6.0, 10.0, 24.0, 10.0)
    );
    
    public PicketFenceBlock(Properties settings) {
        super(settings);

        registerDefaultState(defaultBlockState().setValue(WATERLOGGED, false));
    }

    @Override
    protected void createBlockStateDefinition(StateDefinition.Builder<Block, BlockState> builder) {
        super.createBlockStateDefinition(builder);
        builder.add(SHAPE, FACING, WATERLOGGED);
    }

    @Override
    public BlockState getStateForPlacement(BlockPlaceContext ctx) {
        var state = defaultBlockState().setValue(FACING, ctx.getHorizontalDirection().getOpposite())
            .setValue(WATERLOGGED, ctx.getLevel().getFluidState(ctx.getClickedPos()).getType() == Fluids.WATER);
        return updateShape(ctx.getLevel(), ctx.getClickedPos(), state);
    }

    @Override
    protected BlockState updateShape(BlockState state, LevelReader world, ScheduledTickAccess tickView, BlockPos pos, Direction direction, BlockPos neighborPos, BlockState neighborState, RandomSource random) {
        if (direction.getAxis() == state.getValue(FACING).getAxis()) {
            return updateShape(world, pos, state);
        }

        return state;
    }

    private BlockState updateShape(BlockGetter world, BlockPos pos, BlockState state) {
        var fenceFacing = state.getValue(FACING);
        for (var side : new Direction[] { fenceFacing.getOpposite(), fenceFacing }) {
            var inner = side == fenceFacing;
            var neighborState = world.getBlockState(pos.relative(side));
            var neighborBlock = neighborState.getBlock();
            var neighborFacing = neighborBlock instanceof PicketFenceBlock ? neighborState.getValue(FACING) : null;

            Shape shape;
            if (neighborFacing == fenceFacing.getClockWise()) {
                shape = inner ? Shape.CLOCKWISE_INNER_CORNER : Shape.CLOCKWISE_CORNER;
            } else if (neighborFacing == fenceFacing.getCounterClockWise()) {
                shape = inner ? Shape.COUNTERCLOCKWISE_INNER_CORNER : Shape.COUNTERCLOCKWISE_CORNER;
            } else {
                shape = Shape.STRAIGHT;
            }

            // Prevent funny connections
            if (!(neighborBlock instanceof PicketFenceBlock picketFence) || !picketFence.connectsTo(neighborState, side.getOpposite())) {
                shape = Shape.STRAIGHT;
            }

            if (shape != Shape.STRAIGHT) {
                return state.setValue(SHAPE, shape);
            }
        }

        return state.setValue(SHAPE, Shape.STRAIGHT);
    }

    @Override
    public VoxelShape getShape(BlockState state, BlockGetter world, BlockPos pos, CollisionContext context) {
        return switch (state.getValue(SHAPE)) {
            case STRAIGHT -> STRAIGHT_OUTLINE_SHAPES.get(state.getValue(FACING));
            case CLOCKWISE_CORNER -> CORNER_OUTLINE_SHAPES.get(state.getValue(FACING));
            case COUNTERCLOCKWISE_CORNER -> CORNER_OUTLINE_SHAPES.get(state.getValue(FACING).getCounterClockWise());
            case CLOCKWISE_INNER_CORNER -> CORNER_OUTLINE_SHAPES.get(state.getValue(FACING).getOpposite());
            case COUNTERCLOCKWISE_INNER_CORNER -> CORNER_OUTLINE_SHAPES.get(state.getValue(FACING).getClockWise());
        };
    }

    @Override
    public VoxelShape getCollisionShape(BlockState state, BlockGetter world, BlockPos pos, CollisionContext context) {
        return switch (state.getValue(SHAPE)) {
            case STRAIGHT -> STRAIGHT_COLLISION_SHAPES.get(state.getValue(FACING));
            case CLOCKWISE_CORNER -> CORNER_COLLISION_SHAPES.get(state.getValue(FACING));
            case COUNTERCLOCKWISE_CORNER -> CORNER_COLLISION_SHAPES.get(state.getValue(FACING).getCounterClockWise());
            case CLOCKWISE_INNER_CORNER -> CORNER_COLLISION_SHAPES.get(state.getValue(FACING).getOpposite());
            case COUNTERCLOCKWISE_INNER_CORNER -> CORNER_COLLISION_SHAPES.get(state.getValue(FACING).getClockWise());
        };
    }

    @Override
    public FluidState getFluidState(BlockState state) {
        return state.getValue(WATERLOGGED) ? Fluids.WATER.getSource(false) : super.getFluidState(state);
    }

    @Override
    public BlockState rotate(BlockState state, Rotation rotation) {
        return state.setValue(FACING, rotation.rotate(state.getValue(FACING)));
    }

    @Override
    public BlockState mirror(BlockState state, Mirror mirror) {
        return state.setValue(FACING, mirror.mirror(state.getValue(FACING)));
    }

    @Override
    public boolean isPathfindable(BlockState state, PathComputationType type) {
        return false;
    }

    public boolean sideCoversSmallSquare(BlockState state) {
        return state.getValue(SHAPE) != Shape.STRAIGHT;
    }

    private boolean connectsTo(BlockState state, Direction direction) {
        if (!direction.getAxis().isHorizontal()) return false;

        var facing = state.getValue(FACING);
        return switch (state.getValue(SHAPE)) {
            case STRAIGHT -> facing.getAxis() != direction.getAxis();
            case CLOCKWISE_CORNER -> direction == facing.getCounterClockWise() || direction == facing.getOpposite();
            case COUNTERCLOCKWISE_CORNER -> direction == facing.getClockWise() || direction == facing.getOpposite();
            case CLOCKWISE_INNER_CORNER -> direction == facing.getClockWise() || direction == facing;
            case COUNTERCLOCKWISE_INNER_CORNER -> direction == facing.getCounterClockWise() || direction == facing;
        };
    }

    public enum Shape implements StringRepresentable {
        STRAIGHT("straight"),
        CLOCKWISE_CORNER("clockwise_corner"),
        COUNTERCLOCKWISE_CORNER("counterclockwise_corner"),
        CLOCKWISE_INNER_CORNER("clockwise_inner_corner"),
        COUNTERCLOCKWISE_INNER_CORNER("counterclockwise_inner_corner");

        private final String id;

        Shape(String id) {
            this.id = id;
        }

        @Override
        public String getSerializedName() {
            return id;
        }
    }
}

File: common/src/main/kotlin/juuxel/adorn/block/PicketFenceBlock.kt
@file:Suppress("DEPRECATION")

package juuxel.adorn.block

import juuxel.adorn.util.buildShapeRotationsFromNorth
import juuxel.adorn.util.mergeIntoShapeMap
import juuxel.adorn.util.mergeShapeMaps
import net.minecraft.block.Block
import net.minecraft.block.BlockState
import net.minecraft.block.ShapeContext
import net.minecraft.block.Waterloggable
import net.minecraft.entity.ai.pathing.NavigationType
import net.minecraft.fluid.Fluids
import net.minecraft.item.ItemPlacementContext
import net.minecraft.state.StateManager
import net.minecraft.state.property.BooleanProperty
import net.minecraft.state.property.DirectionProperty
import net.minecraft.state.property.EnumProperty
import net.minecraft.state.property.Properties
import net.minecraft.util.BlockMirror
import net.minecraft.util.BlockRotation
import net.minecraft.util.StringIdentifiable
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Direction
import net.minecraft.util.shape.VoxelShape
import net.minecraft.world.BlockView
import net.minecraft.world.WorldAccess

class PicketFenceBlock(settings: Settings) : Block(settings), Waterloggable, BlockWithDescription {
    init {
        defaultState = defaultState
            .with(FACING, Direction.NORTH)
            .with(SHAPE, Shape.STRAIGHT)
            .with(WATERLOGGED, false)
    }

    override fun appendProperties(builder: StateManager.Builder<Block, BlockState>) {
        super.appendProperties(builder)
        builder.add(SHAPE, FACING, WATERLOGGED)
    }

    override fun getPlacementState(ctx: ItemPlacementContext): BlockState {
        return defaultState.with(FACING, ctx.horizontalPlayerFacing.opposite)
            .with(WATERLOGGED, ctx.world.getFluidState(ctx.blockPos).fluid === Fluids.WATER)
            .let { updateShape(ctx.world, ctx.blockPos, it) }
    }

    override fun getStateForNeighborUpdate(
        state: BlockState, facing: Direction, neighborState: BlockState,
        world: WorldAccess, pos: BlockPos, neighborPos: BlockPos
    ): BlockState {
        if (facing.axis == state[FACING].axis) {
            return updateShape(world, pos, state)
        }

        return state
    }

    private fun updateShape(world: WorldAccess, pos: BlockPos, state: BlockState): BlockState {
        val fenceFacing = state[FACING]
        for (side in arrayOf(fenceFacing.opposite, fenceFacing)) {
            val inner = side == fenceFacing
            val neighborState = world.getBlockState(pos.offset(side))
            val neighborBlock = neighborState.block
            val neighborFacing = if (neighborBlock is PicketFenceBlock) neighborState[FACING] else null

            var shape = when (neighborFacing) {
                fenceFacing.rotateYClockwise() -> {
                    if (inner) {
                        Shape.CLOCKWISE_INNER_CORNER
                    } else {
                        Shape.CLOCKWISE_CORNER
                    }
                }
                fenceFacing.rotateYCounterclockwise() -> {
                    if (inner) {
                        Shape.COUNTERCLOCKWISE_INNER_CORNER
                    } else {
                        Shape.COUNTERCLOCKWISE_CORNER
                    }
                }
                else -> Shape.STRAIGHT
            }

            // Prevent funny connections
            if (neighborBlock !is PicketFenceBlock || !neighborBlock.connectsTo(neighborState, side.opposite)) {
                shape = Shape.STRAIGHT
            }

            if (shape != Shape.STRAIGHT) {
                return state.with(SHAPE, shape)
            }
        }

        return state.with(SHAPE, Shape.STRAIGHT)
    }

    override fun getOutlineShape(state: BlockState, view: BlockView, pos: BlockPos, context: ShapeContext): VoxelShape = when (state[SHAPE]!!) {
        Shape.STRAIGHT -> STRAIGHT_OUTLINE_SHAPES.getValue(state[FACING])
        Shape.CLOCKWISE_CORNER -> CORNER_OUTLINE_SHAPES.getValue(state[FACING])
        Shape.COUNTERCLOCKWISE_CORNER -> CORNER_OUTLINE_SHAPES.getValue(state[FACING].rotateYCounterclockwise())
        Shape.CLOCKWISE_INNER_CORNER -> CORNER_OUTLINE_SHAPES.getValue(state[FACING].opposite)
        Shape.COUNTERCLOCKWISE_INNER_CORNER -> CORNER_OUTLINE_SHAPES.getValue(state[FACING].rotateYClockwise())
    }

    override fun getCollisionShape(state: BlockState, view: BlockView, pos: BlockPos, context: ShapeContext): VoxelShape =
        when (state[SHAPE]!!) {
            Shape.STRAIGHT -> STRAIGHT_COLLISION_SHAPES.getValue(state[FACING])
            Shape.CLOCKWISE_CORNER -> CORNER_COLLISION_SHAPES.getValue(state[FACING])
            Shape.COUNTERCLOCKWISE_CORNER -> CORNER_COLLISION_SHAPES.getValue(state[FACING].rotateYCounterclockwise())
            Shape.CLOCKWISE_INNER_CORNER -> CORNER_COLLISION_SHAPES.getValue(state[FACING].opposite)
            Shape.COUNTERCLOCKWISE_INNER_CORNER -> CORNER_COLLISION_SHAPES.getValue(state[FACING].rotateYClockwise())
        }

    override fun getFluidState(state: BlockState) =
        if (state[WATERLOGGED]) {
            Fluids.WATER.getStill(false)
        } else {
            super.getFluidState(state)
        }

    override fun rotate(state: BlockState, rotation: BlockRotation): BlockState =
        state.with(FACING, rotation.rotate(state[FACING]))

    override fun mirror(state: BlockState, mirror: BlockMirror): BlockState =
        state.with(FACING, mirror.apply(state[FACING]))

    override fun canPathfindThrough(state: BlockState, world: BlockView, pos: BlockPos, type: NavigationType) = false

    fun sideCoversSmallSquare(state: BlockState): Boolean =
        state[SHAPE] != Shape.STRAIGHT

    private fun connectsTo(state: BlockState, direction: Direction): Boolean {
        if (!direction.axis.isHorizontal) {
            return false
        }

        val facing = state[FACING]

        return when (state[SHAPE]!!) {
            Shape.STRAIGHT -> facing.axis != direction.axis
            Shape.CLOCKWISE_CORNER -> direction == facing.rotateYCounterclockwise() || direction == facing.opposite
            Shape.COUNTERCLOCKWISE_CORNER -> direction == facing.rotateYClockwise() || direction == facing.opposite
            Shape.CLOCKWISE_INNER_CORNER -> direction == facing.rotateYClockwise() || direction == facing
            Shape.COUNTERCLOCKWISE_INNER_CORNER -> direction == facing.rotateYCounterclockwise() || direction == facing
        }
    }

    companion object {
        val SHAPE: EnumProperty<Shape> = EnumProperty.of("shape", Shape::class.java)
        val FACING: DirectionProperty = Properties.HORIZONTAL_FACING
        val WATERLOGGED: BooleanProperty = Properties.WATERLOGGED

        private val STRAIGHT_OUTLINE_SHAPES = buildShapeRotationsFromNorth(0, 0, 7, 16, 16, 9)
        private val CORNER_OUTLINE_SHAPES = mergeIntoShapeMap(
            mergeShapeMaps(
                buildShapeRotationsFromNorth(0, 0, 7, 9, 16, 9),
                buildShapeRotationsFromNorth(7, 0, 9, 9, 16, 16)
            ),
            PostBlock.Y_SHAPE
        )
        private val STRAIGHT_COLLISION_SHAPES = buildShapeRotationsFromNorth(0, 0, 7, 16, 24, 9)
        private val CORNER_COLLISION_SHAPES = mergeIntoShapeMap(
            mergeShapeMaps(
                buildShapeRotationsFromNorth(0, 0, 7, 9, 24, 9),
                buildShapeRotationsFromNorth(7, 0, 9, 9, 24, 16)
            ),
            createCuboidShape(6.0, 0.0, 6.0, 10.0, 24.0, 10.0)
        )
    }

    enum class Shape(private val id: String) : StringIdentifiable {
        STRAIGHT("straight"),
        CLOCKWISE_CORNER("clockwise_corner"),
        COUNTERCLOCKWISE_CORNER("counterclockwise_corner"),
        CLOCKWISE_INNER_CORNER("clockwise_inner_corner"),
        COUNTERCLOCKWISE_INNER_CORNER("counterclockwise_inner_corner")
        ;

        override fun asString() = id
    }
}
